### PR TITLE
fatal: specific aarch32 fatal error types

### LINKED ISSUE #53093: ARM: Ability to query CFSR on exception
**Is your enhancement proposal related to a problem? Please describe.**

The CFSR register (Configurable Fault Status Register) is overwritten by the `z_arm_fault` exception handler.
As a result the rich reason information contained in the register is unavailable to the user handler (`k_sys_fatal_error_handler`).
Even the broad class of exception (Usage Fault, Bus Fault, etc) is lost, and the handler only gets `K_ERR_CPU_EXCEPTION`, which is much less useful.
The complete picture of a reboot is only available if logging is enabled and those logs are somehow available.

**Describe the solution you'd like**

A mechanism that preserves the `CFSR` register as it was before modification so user exception handlers can query the original value.

### PR DESCRIPTION
**Pull request by JordanYates**
Define and return more specific fatal error types in aarch32 exception handlers.
Resolves #53093, where specific error information is printed to the terminal but is not available in code.

Fixes #53093

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

### CODE REVIEW

File: tests/kernel/fatal/exception/src/main.c
Comment (JordanYates) [inline tests/kernel/fatal/exception/src/main.c:77]: I am having trouble determining under which conditions `K_ERR_ARM_USAGE_ILLEGAL_EPSR` is generated vs `K_ERR_ARM_MEM_INSTRUCTION_ACCESS`. My hunch is something to do with the MPU (maybe `z_arm_configure_dynamic_mpu_regions`?) Would appreciate if anyone could chime in.

**Comment by microbuilder** [comment]
@JordanYates Thanks for the PR. I definitely want to get this in before the code freeze. Just a heads up that a proper review will likely take me a day or two to get to, since I need to put together and test an update to TF-M that will likely take a the early part of the week for me. Agree the existing code you're amending here isn't ideal.

File: arch/arm/core/aarch32/cortex_m/fault.c
Comment (microbuilder) [inline arch/arm/core/aarch32/cortex_m/fault.c:424]: Nit: 'Imprecise', as per existing SCB definition?

File: include/zephyr/arch/arm/aarch32/arch.h
Comment (microbuilder) [inline include/zephyr/arch/arm/aarch32/arch.h:73]: Imprecise

**Review by microbuilder — CHANGES_REQUESTED** [review]
Minor nit, but appreciate the addition.

**Review by microbuilder — APPROVED** [review]

**Comment by microbuilder** [comment]
@carlocaione @stephanosio Do either of you have time for a second review?

**Review by povergoing — APPROVED** [review]

**Review by carlocaione — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by stephanosio — COMMENTED** [review]
This broke `zephyr/tests/kernel/fatal/no-multithreading/kernel.no-mt.cpu_exception` test on `qemu_cortex_m3`.

```
INFO    - 686/686 qemu_cortex_m3            zephyr/tests/kernel/fatal/no-multithreading/kernel.no-mt.cpu_exception FAILED Failed (qemu 2.011s)
INFO    - /__w/zephyr/zephyr/twister-out/qemu_cortex_m3/zephyr/tests/kernel/fatal/no-multithreading/kernel.no-mt.cpu_exception/handler.log
ERROR   - *** Booting Zephyr OS build zephyr-v3.2.0-3424-g57f5727ac777 ***
Running TESTSUITE fatal_no_mt
===================================================================
START - test_fatal
cpu exception
E: ***** USAGE FAULT *****
E:   Illegal use of the EPSR
E: r0/a1:  0x0000000e  r1/a2:  0x0000000a  r2/a3:  0x90000000
E: r3/a4:  0x20000c3c r12/ip:  0x00000043 r14/lr:  0x000003d5
E:  xpsr:  0x00000000
E: Faulting instruction address (r15/pc): 0x20000c3c
E: >>> ZEPHYR FATAL ERROR 35: Unknown error on CPU 0
E: Current thread: (nil) (unknown)
Caught system error -- reason 35
Unexpected reason (exp: 0)
===================================================================
RunID: 99d50f00015d52359cf40b0ef7c05188
PROJECT EXECUTION FAILED
```

https://github.com/zephyrproject-rtos/zephyr/actions/runs/3901831037/jobs/6665337806#step:13:917

**Comment by microbuilder** [comment]
> This broke `zephyr/tests/kernel/fatal/no-multithreading/kernel.no-mt.cpu_exception` test on `qemu_cortex_m3`.

https://github.com/zephyrproject-rtos/zephyr/pull/53758

**Comment by JordanYates** [comment]
Sorry for missing this and thanks for fixing @microbuilder

**Comment by nashif** [comment]
@JordanYates  This seems to breaking frdm_k64f in  kernel/mem_protect/userspace/ 

```
===================================================================
START - test_pass_user_object
E: 0x20000028 is not a valid k_sem
E: address is not a known kernel object
E: syscall z_vrfy_k_sem_init failed check: access denied
E: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0x00000000
E: r3/a4:  0x00000000 r12/ip:  0x00000000 r14/lr:  0x00000000
E:  xpsr:  0x00000000
E: Faulting instruction address (r15/pc): 0x00000000
E: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0
E: Current thread: 0x20000130 (test_pass_user_object)
Caught system error -- reason 3
System error was expected
 PASS - test_pass_user_object in 0.044 seconds
===================================================================
START - test_read_kernel_data
E: ***** BUS FAULT *****
E:   Precise data bus error
E:   BFAR Address: 0x2000066c
E:   NXP MPU error, port 3
E:     Mode: User, Data Address: 0x2000066c
E:     Type: Read, Master: 0, Regions: 0x8800
E: r0/a1:  0x00010ef2  r1/a2:  0x00000002  r2/a3:  0x00000001
E: r3/a4:  0x2000066c r12/ip:  0x0000bc81 r14/lr:  0x00000cb9
E:  xpsr:  0x21000000
E: Faulting instruction address (r15/pc): 0x00000cbc
E: >>> ZEPHYR FATAL ERROR 25: Unknown error on CPU 0
E: Current thread: 0x20000130 (test_read_kernel_data)
Caught system error -- reason 25
Wrong fault reason, expecting 19
```

We are opening a blocker bugs for this, frdm_k64 is one of thoe platforms we need to have all key kernel tests passing.

@hakehuang  @dleach02  FYI

**Comment by hakehuang** [comment]
Thansk anas, I report this issue https://github.com/zephyrproject-rtos/zephyr/issues/54053

**Comment by JordanYates** [comment]
Without one of those boards on hand I can't really validate that any fixes I make actually fix all the tests

**Pull request merged**
